src/main/kotlin/io/snyk/plugin/ui/toolwindow/SnykToolWindowSnykCodeScanListenerLS.kt

```diff
@@ -0,0 +1,166 @@
+import io.snyk.plugin.ui.toolwindow.nodes.secondlevel.SnykCodeFileTreeNodeFromLS
+import snyk.common.ProductType
+import snyk.common.lsp.ScanIssue
+import snyk.common.lsp.SnykScanParams
+import javax.swing.JTree
+import javax.swing.tree.DefaultMutableTreeNode
+
+class SnykToolWindowSnykCodeScanListenerLS(
+    val project: Project,
+    private val snykToolWindowPanel: SnykToolWindowPanel,
+    private val vulnerabilitiesTree: JTree,
+    private val rootSecurityIssuesTreeNode: DefaultMutableTreeNode,
+    private val rootQualityIssuesTreeNode: DefaultMutableTreeNode,
+) : SnykCodeScanListenerLS {
+
+    override fun scanningStarted(snykScan: SnykScanParams) {
+        ApplicationManager.getApplication().invokeLater {
+            rootSecurityIssuesTreeNode.userObject = "$CODE_SECURITY_ROOT_TEXT (scanning...)"
+            rootQualityIssuesTreeNode.userObject = "$CODE_QUALITY_ROOT_TEXT (scanning...)"
+        }
+    }
+
+    override fun scanningSnykCodeFinished(snykCodeResults: Map<SnykCodeFile, List<ScanIssue>>) {
+        ApplicationManager.getApplication().invokeLater {
+            displaySnykCodeResults(snykCodeResults)
+            refreshAnnotationsForOpenFiles(project)
+        }
+    }
+
+    override fun scanningSnykCodeError(snykScan: SnykScanParams) = Unit
+
+    fun displaySnykCodeResults(snykCodeResults: Map<SnykCodeFile, List<ScanIssue>>) {
+        if (getSnykCachedResults(project)?.currentSnykCodeError != null) return
+        if (pluginSettings().token.isNullOrEmpty()) {
+            snykToolWindowPanel.displayAuthPanel()
+            return
+        }
+        val selectedNodeUserObject = TreeUtil.findObjectInPath(vulnerabilitiesTree.selectionPath, Any::class.java)
+
+        // display Security issues
+        val userObjectsForExpandedSecurityNodes =
+            snykToolWindowPanel.userObjectsForExpandedNodes(rootSecurityIssuesTreeNode)
+        rootSecurityIssuesTreeNode.removeAllChildren()
+
+        var securityIssuesCount: Int? = null
+        var securityIssuesHMLPostfix = ""
+        if (pluginSettings().snykCodeSecurityIssuesScanEnable) {
+            val securityResults = snykCodeResults
+                .map { it.key to it.value.filter { issue -> issue.additionalData.isSecurityType } }
+                .toMap()
+            securityIssuesCount = securityResults.values.flatten().distinct().size
+            securityIssuesHMLPostfix = buildHMLpostfix(securityResults)
+
+            if (pluginSettings().treeFiltering.codeSecurityResults) {
+                val securityResultsToDisplay = securityResults.map { entry ->
+                    entry.key to entry.value
+                        .filter { pluginSettings().hasSeverityEnabledAndFiltered(it.getSeverityAsEnum()) }
+                }.toMap()
+
+                displayResultsForCodeRoot(
+                    rootSecurityIssuesTreeNode,
+                    securityResultsToDisplay
+                )
+            }
```

Comment:
In the description of the PR you commented that there are a few duplications in the code. Is this one highlighted by the IDE one of them or we could encapsulate this in a method?

<img width="1684" alt="image" src="https://github.com/snyk/snyk-intellij-plugin/assets/1948377/69b59cb9-772a-4ff5-b76e-2874ac1ae223">

Reply:
Will have a go at cleaning this up tomorrow.